test/create_db.sh

```diff
@@ -16,134 +26,82 @@ function print_heading() {
-  migrations=($(get_migrations "${db_mig_path}"))
-
-  # Goose up, this will work if there are schema files present at
-  # $dbpath with a newer timestamp than the current goose dbversion.
-  apply_migrations "${migrations}" "${dbpath}" "${dbenv}" "${db}"
-
-  # The (actual) latest migration should always be the last file or
-  # symlink at $db_mig_path.
-  latest_mig_path_filename="$(basename -- "${migrations[-1]}")"
-
-  # Goose's dbversion is the timestamp (first 14 characters) of the file
-  # that it last migrated to. We can figure out which goose dbversion we
-  # should be on by parsing the timestamp of the latest file at
-  # $db_mig_path.
-  latest_db_mig_version="${latest_mig_path_filename:0:14}"
-  
-  # Ask Goose the timestamp (dbversion) our database is currently
-  # migrated to.
-  goose_dbversion="$(goose -path=${dbpath} -env=${dbenv} dbversion | sed 's/goose: dbversion //')"
-
-  # If the $goose_dbversion does not match the $latest_in_db_mig_path,
-  # trigger recreate
-  if [[ "${latest_db_mig_version}" != "${goose_dbversion}" ]]; then
-    print_heading "Detected latest migration version mismatch"
-    echo "dropping and recreating from migrations at ${db_mig_path}"
-    create_empty_db "${db}" "${dbconn}"
-    apply_migrations "${migrations}" "${dbpath}" "${dbenv}" "${db}"
-  fi
-
-  # With MYSQL_CONTAINER, patch the GRANT statements to
-  # use 127.0.0.1, not localhost, as MySQL may interpret
-  # 'username'@'localhost' to mean only users for UNIX
-  # socket connections. Use '-f' to ignore errors while
-  # we have migrations that haven't been applied but
-  # add new tables (TODO(#2931): remove -f).
-  USERS_SQL=test/sa_db_users.sql
-  if [[ ${MYSQL_CONTAINER} ]]; then
-    sed -e "s/'localhost'/'%'/g" < ${USERS_SQL} | \
-      mysql $dbconn -D $db -f || exit_err "unable to add users to ${db}"
-  else
-    sed -e "s/'localhost'/'127.%'/g" < $USERS_SQL | \
-      mysql $dbconn -D $db -f < $USERS_SQL || exit_err "unable to add users to ${db}"
-  fi
-  echo "added users to ${db}"
+mysql ${dbconn} -e "SET GLOBAL max_connections = 500;"
+
+for db in $DBS; do
+  for env in $ENVS; do
+    dbname="${db}_${env}"
+    print_heading "${dbname}"
+    if mysql ${dbconn} -e 'show databases;' | grep "${dbname}" > /dev/null; then
+      echo "Already exists - skipping create"
+    else
+      echo "Doesn't exist - creating"
+      create_empty_db "${dbname}" "${dbconn}"
+    fi
+
+    if [[ "${BOULDER_CONFIG_DIR}" == "test/config-next" ]]
+    then
+      dbpath="./sa/db-next"
+    else
+      dbpath="./sa/db"
+    fi
+
+    # sql-migrate will default to ./dbconfig.yml and treat all configured dirs
+    # as relative.
+    cd "${dbpath}"
+    r=`sql-migrate up -env="${dbname}" | xargs echo`
+    if [[ "${r}" == "Migration failed"* ]]
+    then
+      echo "Migration failed - dropping and recreating"
+      create_empty_db "${dbname}" "${dbconn}"
+      sql-migrate up -env="${dbname}"
```

Comment:
When a migration fails, I think it would be better to tell the user (and give them instructions on how to drop the database) rather than automatically drop the database.

Reply:
I'll make this a little clearer (to the end user) with a small change, but if the migration fails after the drop and re-apply it will also spit out the error and exit.

Reply:
Aha, you're right! Thanks for explaining that.